src/Products/PluginIndexes/KeywordIndex/KeywordIndex.py

```diff
@@ -57,63 +64,113 @@ def _index_object(self, documentId, obj, threshold=None, attr=''):
-            try:
-                for kw in newKeywords:
-                    self.insertForwardIndexEntry(kw, documentId)
-                if newKeywords:
-                    self._unindex[documentId] = list(newKeywords)
-            except TypeError:
-                return 0
+            if self.providesSpecialIndex(newKeywords):
+                self.insertSpecialIndexEntry(newKeywords, documentId)
+            else:
+                try:
+                    self.index_objectKeywords(documentId, newKeywords)
+                except self.exceptions_treated_as_missing:
+                    LOG.error('%(context)s: Unable to insert forward '
+                              'index entry for document with id '
+                              '%(doc_id)s and keywords %(keywords)r '
+                              'for index %{index}r.', dict(
+                                  context=self.__class__.__name__,
+                                  keywords=newKeywords,
+                                  doc_id=documentId,
+                                  index=self.id))
+                    if self.providesSpecialIndex(missing):
+                        newKeywords = missing
+                        self.insertSpecialIndexEntry(missing, documentId)
+                    else:
+                        return 0
         else:
             # we have an existing entry for this document, and we need
             # to figure out if any of the keywords have actually changed
-            if type(oldKeywords) is not OOSet:
-                oldKeywords = OOSet(oldKeywords)
-            newKeywords = OOSet(newKeywords)
-            fdiff = difference(oldKeywords, newKeywords)
-            rdiff = difference(newKeywords, oldKeywords)
-            if fdiff or rdiff:
-                # if we've got forward or reverse changes
-                if newKeywords:
-                    self._unindex[documentId] = list(newKeywords)
+            if self.providesSpecialIndex(oldKeywords) and \
+               self.providesSpecialIndex(newKeywords):
+                if oldKeywords == newKeywords:
+                    return 0
+                self.removeSpecialIndexEntry(oldKeywords, documentId)
+                self.insertSpecialIndexEntry(newKeywords, documentId)
+                self._unindex[documentId] = newKeywords
+                return 1
+
+            if self.providesSpecialIndex(oldKeywords):
+                self.removeSpecialIndexEntry(oldKeywords, documentId)
+                oldSet = OOSet()
+            else:
+                if not isinstance(oldKeywords, OOSet):
+                    oldKeywords = OOSet(oldKeywords)
+                oldSet = oldKeywords
+
+            if self.providesSpecialIndex(newKeywords):
+                self.insertSpecialIndexEntry(newKeywords, documentId)
+                newSet = OOSet()
+            else:
+                newSet = newKeywords = OOSet(newKeywords)
+
+            try:
+                fdiff = difference(oldSet, newSet)
```

Comment:
This will fail if the keywords change type - and will let your index in a strange state. Assume "oldSet" to be `OOSet(['a', 'b']), "newSet" to be `OOSet([1, 2])`. Then under Python3, the `difference` will result in a `TypeError` which lets your document remain indexed under `oldSet` and gets newly indexed under `missing`.

Under Python 3, all indexed values must have a common type. Changing the keyword type will therefore not work (apart from constructed cases). Therefore, you should let an exception from the `difference` calls propagate (maybe log and provide a more specific error message) and not turn it into `missing`.